File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    private List<List<Integer>> result;

    private void backTrack(int[] candidates, int target, int curr, int sum, LinkedList<Integer> currRes) {
        if (sum > target) {
            return;
        }
        if (sum == target) {
            result.add(new LinkedList<>(currRes));
            return;
        }
        currRes.addLast(candidates[curr]);
        backTrack(candidates, target, curr, sum + candidates[curr], currRes);
        currRes.removeLast();
        if (curr < candidates.length - 1) {
            backTrack(candidates, target, curr+1, sum, currRes);
        }
    }

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        Arrays.sort(candidates);
        this.result = new LinkedList<>();
        this.backTrack(candidates, target, 0, 0, new LinkedList<>());
        return this.result;
    }
}

File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    def __init__(self):
        self.result = []

    def dfs(self, curr, candidates, target, currSum, currList):
        print(currSum, currList)
        if currSum > target:
            return
        if currSum == target:
            res = currList.copy()
            self.result.append(res)
            return
        for i in range(curr, len(candidates)):
            currList.append(candidates[i])
            self.dfs(i, candidates, target, currSum + candidates[i], currList)
            currList.pop()
        return

    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        self.dfs(0, candidates, target, 0, [])
        return self.result